packages/mds-agency/api.ts

```diff
@@ -316,73 +316,51 @@ function api(app: express.Express): express.Express {
-          telemetry: null,
-          timestamp: recorded,
-          trip_id: null,
-          recorded,
-          telemetry_timestamp: undefined,
-          service_area_id: null
-        }
-        db.writeEvent(event)
-          .then(() => {
-            Promise.all([cache.writeEvent(event), stream.writeEvent(event)])
-              .then(resolve)
-              .catch(err => /* istanbul ignore next */ {
-                log.warn('/event exception cache/stream', err)
-                reject()
-              })
-          }, reject)
-          .catch(reject)
-      })
-    }
-
-    function success(): void {
-      device.status = VEHICLE_STATUSES.removed
-      log.info('new', providerName(res.locals.provider_id), 'vehicle added', JSON.stringify(device)).then(() => {
-        writeRegisterEvent().then(
-          () => {
-            res.status(201).send({
-              result: 'register device success',
-              recorded,
-              device
-            })
-          },
-          err => /* istanbul ignore next */ {
-            log.error('register device failed to write register event', err).then(() => {
-              res.status(500).send({
-                result: 'register device internal error'
-              })
-            })
-          }
-        )
-      })
+      const event: VehicleEvent = {
+        device_id: device.device_id,
+        provider_id: device.provider_id,
+        event_type: VEHICLE_EVENTS.register,
+        event_type_reason: null,
+        telemetry: null,
+        timestamp: recorded,
+        trip_id: null,
+        recorded,
+        telemetry_timestamp: undefined,
+        service_area_id: null
+      }
+      try {
+        await db.writeEvent(event)
+        await Promise.all([cache.writeEvent(event), stream.writeEvent(event)]).catch(async err => {
+          await log.warn('/event exception cache/stream', err)
+        })
+      } catch (err) {
+        await log.error(err)
+        throw new Error('writeEvent exception db')
+      }
     }
 
     // writing to the DB is the crucial part.  other failures should be noted as bugs but tolerated
     // and fixed later.
-    db.writeDevice(device).then(
-      () => {
-        Promise.all([cache.writeDevice(device), stream.writeDevice(device)]).then(success, success)
-      },
-      (err: Error | string) => /* istanbul ignore next */ {
-        if (String(err).includes('duplicate')) {
-          res.status(409).send({
-            error: 'already_registered',
-            error_description: 'A vehicle with this device_id is already registered'
-          })
-        } else {
-          log.error(providerName(res.locals.provider_id), 'register vehicle failed:', err).then(() => {
-            res.status(500).send(new ServerError())
-          })
-        }
+    try {
+      await db.writeDevice(device)
+      await Promise.all([cache.writeDevice(device), stream.writeDevice(device)])
+      device.status = VEHICLE_STATUSES.removed
```

Comment:
The device has already been written to db/cache/stream .. what does setting the status here accomplish? In general, I would prefer `res.status(201).send({ result: 'register device success', recorded, device: { ...device, status: VEHICLE_STATUSES.removed })`

Reply:
probably that mutation should happen at the top